tests/conftest.py

```diff
@@ -43,14 +44,34 @@ def get_database_url():
 
 
 @pytest.fixture(scope="session")
-def db_engine():
+def db_engine(tmp_path_factory):
     engine = sqlalchemy.create_engine(TEST_SETTINGS["database_url"])
 
-    # Delete all database tables and re-initialize the database schema based on
-    # the current models. Doing this at the beginning of each test run ensures
-    # that any schema changes made to the models since the last test run will
-    # be applied to the test DB schema before running the tests again.
-    db.init(engine, drop=True, stamp=False)
+    # Use a filelock to only init the DB once even though we have multiple
+    # parallel pytest-xdist workers. See:
+    # https://pytest-xdist.readthedocs.io/en/stable/how-to.html?highlight=filelock#making-session-scoped-fixtures-execute-only-once
+
+    # The temporary directory shared by all pytest-xdist workers.
+    shared_tmpdir = tmp_path_factory.getbasetemp().parent
+
+    # The existence of this file records that a worker has initialized the DB.
+    done_file = shared_tmpdir / "db_initialized"
+
+    with FileLock(str(done_file) + ".lock"):
+        if done_file.is_file():
+            # Another worker already initialized the DB.
+            pass
+        else:
+            # Delete all database tables and re-initialize the database schema
+            # based on the current models. Doing this at the beginning of each
+            # test run ensures that any schema changes made to the models since
+            # the last test run will be applied to the test DB schema before
+            # running the tests again.
+            db.init(engine, drop=True, stamp=False)
+
+            # Make sure that no other worker tries to init the DB after we
+            # release the lock file.
+            done_file.touch()
```

Comment:
Not sure I understand this part and/or the comment.

I this leaving the lock file around independently of the context manager?

Reply:
The lock file itself isn't left around outside of the `with`, but it's necessary to leave a separate "done file" around. There are two files: the "done file" (called just `db_initialized`) and the lock file (`db_initialized.lock`). What happens is:

1. One of the test processes reaches `with FileLock(str(done_file) + ".lock")` and acquires the lock, initializes the DB, and creates the done file
2. The second test process reaches the `with FileLock(str(done_file) + ".lock")` line and blocks, waiting to acquire the lock file. When the first process exits the `with` the second process acquires the lock file and enters the `with` but now, because the first process has created the "done file", the second process doesn't try to initialize the DB again a second time
3. The rest of the test processes proceed the same way as the second process did

The entire `shared_tmpdir` (containing the done file) is unique to and deleted after each test session, so separate test sessions don't affect each other.

Reply:
With pytest-xdist, session-scoped fixtures like `db_engine` get executed multiple times: once for each parallel process that the tests have been divided between (one process per core). It makes sense because they're separate process, so each one is going to need its own separate DB engine object. But they would all try to run the `db_engine` fixture and all try to initialize the DB, hence why this locking mechanism is needed

Reply:
Tweaked the code a bit to hopefully make this a bit clearer